src/pki/commands/ocsp.c

```diff
@@ -0,0 +1,569 @@
+		response->serialNumber   = chunk_clone(serialNumber);
+		response->thisUpdate     = time(NULL);
+		DBG1(DBG_LIB, "  thisUpdate:        %#T", &response->thisUpdate, TRUE);
+
+		if (lifetime)
+		{
+			response->nextUpdate = response->thisUpdate + lifetime;
+			DBG1(DBG_LIB, "  nextUpdate:        %#T", &response->nextUpdate, TRUE);
+		}
+
+		if (issuer_cacert && (issuer_cacert == first_issuer || self_signed))
+		{
+			status = ocsp_responder->get_status(ocsp_responder,
+									issuer_cacert,	serialNumber,
+									&revocationTime, &revocationReason);
+		}
+		DBG1(DBG_LIB, "  certValidation:    %N", cert_validation_names, status);
+		response->status = status;
+
+		if (status == VALIDATION_REVOKED || status == VALIDATION_ON_HOLD)
+		{
+			DBG1(DBG_LIB, "  revocationTime:    %T", &revocationTime, TRUE);
+			DBG1(DBG_LIB, "  revocationReason:  %N", crl_reason_names,
+												  revocationReason);
+			response->revocationTime   = revocationTime;
+			response->revocationReason = revocationReason;
+		}
+		responses->insert_last(responses, response);
+	}
+	enumerator->destroy(enumerator);
+
+	if (multiple_issuers)
+	{
+		DBG1(DBG_LIB, "there are multiple known issuers");
+	}
+
+gen:
+	if (op == OP_RESPOND)
+	{
+		if (cert)
+		{
+			DBG1(DBG_LIB, "%s \"%Y\"",
+				 self_signed ? "self-signed signer:" : "trusted signer:    ",
+				 cert->get_subject(cert));
+
+			scheme = get_signature_scheme(key, digest, pss);
+			if (scheme)
+			{
+				if (digest == HASH_UNKNOWN)
+				{
+					digest = hasher_from_signature_scheme(scheme->scheme,
+														  scheme->params);
+				}
+			}
+			else
+			{
+				DBG1(DBG_LIB, "no signature scheme found");
+				ocsp_status = OCSP_INTERNALERROR;
+			}
+		}
+		DBG1(DBG_LIB, "ocspResponseStatus:  %N", ocsp_status_names, ocsp_status);
+
+		enumerator = responses->create_enumerator(responses);
+		ocsp_resp = lib->creds->create(lib->creds, CRED_CERTIFICATE,
+						CERT_X509_OCSP_RESPONSE,
+						BUILD_OCSP_STATUS, ocsp_status,
+						BUILD_OCSP_RESPONSES, enumerator,
+						BUILD_SIGNING_KEY, key,
+						BUILD_SIGNING_CERT, cert,
+						BUILD_SIGNATURE_SCHEME, scheme,
+						BUILD_NONCE, nonce,
+						BUILD_END);
+		enumerator->destroy(enumerator);
+
+		if (!ocsp_resp)
+		{
+			error = "generating OCSP response failed";
+			goto end;
+		}
+		if (!ocsp_resp->get_encoding(ocsp_resp, form, &encoding))
+		{
+			error = "encoding OCSP response failed";
+			goto end;
+		}
+		set_file_mode(stdout, form);
+		if (fwrite(encoding.ptr, encoding.len, 1, stdout) != 1)
+		{
+			error = "writing OCSP response failed";
+			goto end;
+		}
+	}
+	res = 0;
+
+end:
+	DESTROY_IF(key);
+	lib->credmgr->remove_local_set(lib->credmgr, &creds->set);
+	creds->destroy(creds);
+	responses->destroy_offset(responses,
+							  offsetof(ocsp_single_response_t, destroy));
+	DESTROY_IF(ocsp_req);
+	DESTROY_IF(ocsp_resp);
+	signature_params_destroy(scheme);
+	free(encoding.ptr);
+	if (error)
+	{
+		fprintf(stderr, "%s\n", error);
```

Comment:
Use `DBG1()` here as well instead of `fprintf()` directly? Or maybe use `DBG1()` explicitly for some of the errors above when generating the response? (I assume you used `DBG1()` so that output could - in theory, as `pki`  currently doesn't support any loggers other than the default to `stderr` -  be redirected when the command is called as a service. If so, some of the errors that could occur during generation wouldn't end up there.)

Reply:
Changed DBG_LIB to DBG_APP